### ISSUE TITLE
Getting the smallest prime factors from a FactorSieve

### ISSUE THREAD

**Issue by ExpHP** [cartazio/arithmoi#34]
It does not seem that there is an inexpensive way to look up the contents of a `FactorSieve`, which is unfortunate because a factor sieve can have other useful purposes beyond generating prime factorizations.

Due to memory optimizations I imagine that directly exposing the container is out of the question - but even just a lookup function like `sieveSmallestFactor :: Sieve -> Integer -> Integer` (or `Maybe Integer`) would be immensely useful.

Of course, there are the edge cases of how to handle inputs of `0` or `1`.

**Comment by Bodigrim** [comment]
Sounds reasonable. Could you please prepare a PR?

**Comment by ExpHP** [comment]
Okay, so I looked into this yesterday, and at first wrote a rather conservative version (error unless `2 <= x && x <= bound`).  Then I discovered how `fsIsPrime` is defined, and after further consideration it seems that the definition which fits best into the existing API is as follows:

(written here in slightly pseudocode-like fashion to emphasize all of the edge cases)

```haskell
fsSmallestFactor :: FactorSieve -> Integer -> Integer`
fsSmallestFactor fs x
  |      x <  0      = fsSmallestFactor fs (-x)
  |      x == 0      = 2
  |      x == 1      = error "one has no prime divisors"
  | 2 <= x <= bound  = (fsLookup fs) x
  |      x >  bound  = useSomeFallbackStrategy x
```

The result for `0` may seem unusual at first, but this has the mathematically simple description that the output is the least prime `p` (or equivalently the least integer > 1) which divides evenly into `x`.  Notice this definition has a single error case, so the function can be made total in an unambiguous fashion by changing the return type to `Maybe Integer`.

---

I am working on a PR, but it is going at a somewhat leisurely pace as this is my first experience using cabal for anything beyond a toy script.  I am also not familiar at all with QuickCheck (or the notion of automated, randomized tests in general).

On that note: Am I correct in my assessment that `Math.NumberTheory.Primes.Sieve.Misc.fsPrimeTest` is not publicly exposed? (I am not sure how to easily search the exposed surface API of a haskell package beyond doing a lot of `grep -r`.)

**Comment by Bodigrim** [comment]
Thanks for your effort. No hurry :)

`fsPrimeTest` is in the export list of `Math.NumberTheory.Primes.Sieve.Misc`, but `Math.NumberTheory.Primes.Sieve.Misc` is not amongst exposed modules in `arithmoi.cabal`. This means that you can freely use `fsPrimeTest` in other modules of `arithmoi`, but you cannot run it from other packages. 

One can add `fsPrimeTest` to the export list of `Math.NumberTheory.Primes.Sieve`. This module is exposed, so the function will become publicly available outside of the package.

**Comment by Bodigrim** [comment]
@ExpHP any progress on a PR?

**Comment by ExpHP** [comment]
This fell through the cracks.  Status:

* I don't currently need this in any of my ongoing projects
* I made a branch with the feature about a year ago.  Found it and rebased it without trouble.
* Having had some experience now with Haskell's test frameworks, I was able to add a test this morning.

so I was hoping to get at least something up today...

but I'm not actually able to run the tests due to some [lovely linker errors](https://gist.github.com/ExpHP/0e16f3261dd4f8763ff99515af2c4341) that I've not yet been able to resolve; the packages in the error simply never get installed by cabal into the sandbox.  (Perhaps not coincidentally, it appears that those same packages are all ones that have a copy installed through `pacman`...).  In summary: *sigh...*

I might take another shot at it this weekend

**Comment by cartazio** [comment]
try cabal install --enable-tests --only-deps :)

**Comment by cartazio** [comment]
or dont use the arch packaged ones, possibly both!

**Comment by cartazio** [comment]
i recommend using cabal 2.0 and new build :)

**Comment by ExpHP** [comment]
Ah yes, I forgot about `new-build`!  I used to use that all the time. I still have the script I used to use to locate the binaries since back when I used it there still was no `new-run`.

...unfortunately, `new-build` eventually results in the same linker errors.
...and `stack init` complains that I'm not on cabal 2.0.  (arch has 1.24)
...and `cabal install cabal-install` results in, you guessed it, linker errors!

Reading the Arch wiki, *they did change things!* (emphasis mine):

> GHC uses static linking by default and the -dynamic flag is needed to select dynamic linking. Since version 8.0.2-1, **the Arch ghc package no longer contains static versions of the GHC boot libraries by default,** nor do any of the `haskell-*` packages . Static versions of the GHC boot libraries may be installed separately through the ghc-static package, but no such equivalent exists for the haskell-* packages. Therefore, without -dynamic Haskell code and packages will generally fail to link unless the program depends only on boot packages and locally installed packages and ghc-static is installed.
> 
> This also causes issues with Cabal trying to use the default static linking. To force dynamic linking in Cabal, edit ~/.cabal/config and add the line executable-dynamic: True.
> 
> Dynamic linking is used for most Haskell modules packaged through pacman and is common for packages in the AUR. Since GHC provides no ABI compatibility between compiler releases, static linking is often the preferred option for local development outside of the package system.

So I tried a number of increasingly drastic solutions, but in the end found that I was unable to ever get dynamic linking to work.  Even with `executable-dynamic: True` and all Arch packages uninstalled except `ghc` and `cabal-install`, I would get linker errors if I tried to `cabal install cabal-install` (with or without `--enable-shared`).

So... back to static it is.

```bash
sudo pacman -R --recursive ghc cabal-install xmonad stack idris hlint ghc-mod cabal-helper
sudo updatedb && locate libHS  # good, nothing.
sudo pacman -S ghc ghc-static
(cd ~/asd/clone/cabal && rm -rf * && git reset --hard && cd cabal-install && ./bootstrap.sh)
cabal update
cd ~/asd/clone/arithmoi
cabal new-test
```

...and there we go; I'm back in action. :D

**Comment by ExpHP** [comment]
So... I benchmarked my PR against the fastest alternative, and: **I'm not impressed.**

[Gist](https://gist.github.com/ExpHP/026c61100f3b913c09d429dd04a876dd)

I tried making a sieve of size 2e8 and summing it up.

* `fst . head . sieveFactor sieve` - takes about 30s
* `fromJust . sieveSmallestFactor sieve` - takes about 15s.

A 50% speedup is alright, but I was hoping for orders of magnitude.  My thoughts on why the winnings are so meager:

- Looking more closely, I think `sieveFactor` is already tail recursive modulo cons, so it only needs to strip off powers of the smallest prime.
- sieveSmallestFactor handles lots of edge cases because it must accept arbitrary integers

I am beginning to doubt that it's possible to achieve a speedup of the magnitude I was hoping for with any (safe) function that has the signature `FactorSieve -> Integer -> Integer`.  We would have to return some form of sequence, or some kind of interface for folding over smallest factors.
